fog/view/enclave/impl/src/lib.rs

```diff
@@ -359,34 +397,155 @@ where
+        assert_eq!(
+            result.highest_processed_block_count,
+            last_response.query_response.highest_processed_block_count
+        );
+        assert_eq!(
+            result.highest_processed_block_signature_timestamp,
+            last_response
+                .query_response
+                .highest_processed_block_signature_timestamp
+        );
+        assert_eq!(
+            result.last_known_block_count,
+            last_response.query_response.last_known_block_count
+        );
+    }
+
+    #[test]
+    fn multiple_incomplete_responses_returns_response_before_first_incomplete() {
+        const STORE_COUNT: usize = 3;
+        let mut decrypted_query_responses = Vec::with_capacity(STORE_COUNT);
+
+        // Make the first response fully processed.
+        let first_query_response = create_query_response(1, 0, 1);
+        let block_range = BlockRange::new(0, 1);
+        let decrypted_query_response = DecryptedMultiViewStoreQueryResponse {
+            query_response: first_query_response.clone(),
+            block_range,
+        };
+        decrypted_query_responses.push(decrypted_query_response);
+
+        // Make the second response "incomplete"- i.e. it hasn't processed all of its
+        // blocks.
+        let incomplete_block_count = 0;
+        let incomplete_query_response =
+            create_query_response(incomplete_block_count, 2, incomplete_block_count);
+        let block_range = BlockRange::new(1, 2);
+        decrypted_query_responses.push(DecryptedMultiViewStoreQueryResponse {
+            query_response: incomplete_query_response,
+            block_range,
+        });
+
+        // Make the third response fully processed.
+        let fully_processed_block_count = 3;
+        let query_response =
+            create_query_response(fully_processed_block_count, 3, fully_processed_block_count);
+        let block_range = BlockRange::new(2, 3);
+        decrypted_query_responses.push(DecryptedMultiViewStoreQueryResponse {
+            query_response,
+            block_range,
+        });
+
+        // Make the fourth response incomplete.
+        let block_count = 0;
+        let query_response = create_query_response(block_count, 0, block_count);
+        let block_range = BlockRange::new(3, 4);
+        decrypted_query_responses.push(DecryptedMultiViewStoreQueryResponse {
+            query_response,
+            block_range,
+        });
+
+        let result = get_block_data(decrypted_query_responses.clone());
+
+        assert_eq!(
+            result.highest_processed_block_count,
+            first_query_response.highest_processed_block_count
+        );
+        assert_eq!(
+            result.highest_processed_block_signature_timestamp,
+            first_query_response.highest_processed_block_signature_timestamp
+        );
+        assert_eq!(
+            result.last_known_block_count,
+            first_query_response.last_known_block_count
+        );
```

Comment:
:wrench: Might consider something like the suggestion below
It requires adding `yare` to the Cargo.toml as well as deriving `Debug` and `PartialEq` on the `BlockData`
When doing a loop with an early return one should test; 0, the first, a middle element, the last element, exiting the loop, the current tests are only looking at going all the way through the loop and a middle element.  Adding test cases for the other 2 will probably blow this file up.
I'm not fond of the tuple for the input parameters to the test function, but putting a type there would have put more noise in the array/slice.
Since the time and and last known block are just passthroughs, I used an offset to make sure they copied correctly.

:question:It does bring up a question about `last_known` should we find the highest of that one?  I'm guessing the `last_known` should be pretty close for most shards so probably won't matter.

```suggestion
    #[test]
    fn no_responses() {
        let decrypted_query_responses = vec![];
        assert_eq!(get_block_data(decrypted_query_responses), BlockData::new(0, 0, 0));
    }

    // The tuple is (block_start, block_end, highest_processed)
    #[parameterized(
    all = { &[(0, 4, 4), (5, 8, 8), (9, 11, 11)], 2 },
    first_partial = { &[(0, 4, 3), (5, 8, 8), (9, 11, 11)], 0 },
    middle_partial = { &[(0, 4, 4), (5, 8, 7), (9, 11, 11)], 1 },
    last_partial = { &[(0, 4, 4), (5, 8, 8), (9, 11, 10)], 2 },
    )]
    fn block_data(block_data: &[(u64, u64, u64)], expected_index: usize) {
        let time_shift = 3;
        let known_shift = 2;
        let mut decrypted_query_responses = Vec::with_capacity(block_data.len());
        decrypted_query_responses.extend(block_data.into_iter().map(|b| {
            let block_range = BlockRange::new(b.0, b.1);
            let query_response = create_query_response(b.2, b.2 + time_shift, b.2 + known_shift);
            DecryptedMultiViewStoreQueryResponse {
                query_response,
                block_range,
            }
        }));

        let data = block_data[expected_index];
        let expected_block = BlockData::new(data.2, data.2 + time_shift, data.2 + known_shift);
        assert_eq!(get_block_data(decrypted_query_responses), expected_block);
```

Reply:
While the parametrized version is more concise, I find it unreadable. The the tuple arguments to the `#[parametrized]]` block and devices such as `time_shift` and `known_shift` are hard to follow. I tend to prefer keeping test set up declarative so that you don't have to do much mental gymnastics to see what's going on. 

I am ok with adding those additional test cases, but I think that every case is covered (early return and non-early return). Not sure we have this 0, first ,middle, and last requirement for unit tests elsewhere, but if you feel strongly I will add them.

Reply:
And I think it makes sense to return the highest `last_known_block_index` so I  changed that.

Reply:
- parameterized vs not. I can understand the concerns especially in this instance. I didn't like the array of tuples but couldn't think of a clearer approach. I would encourage keeping the tool availability in ones mind as I feel there are is a large amount of workflow tests in the codebase. I think having people see parameterized tests in the actual codebase will make others consider it over some of the workflow style tests.

. "every case covered"  This is one of the gotcha's of code coverage.  Code coverage is  negative metric, meaning if you don't have it, you know you didn't meet some metric so you know you for sure didn't meet the higher metric of "sufficent" tests.  Having code coverage means that you're somewhere between the code coverage bar and "sufficient" tests bar, but you don't know where.  I quoted sufficient as it is very subjective and there are strong debates about code coverage vs not.
Considering testing coverage, for searching an iterable, one should have:
   - empty iterable, nothing found, this ensures the default value is correct. What if the logic maxed out and had some kind of state of initialized with item from iterable.
   - only one item, this is the boundary for 0 items.  It also ensures someone doesn't have a stray next() call
   - first item, this ensures one doesn't return prior to copying the item value.  Could probably be omitted with a one item test.
   - middle item, this ensures subsequent items don't overwrite it
   - last item, this ensures it's copied correctly and prevents bailing if there is some kind of current <=> next comparison happening
   
I'm coming from C so it's possible there are some language constructs that prevent these so they can be omitted...
Also this is me coming from an opaque box perspective and other times I will argue it's OK to simplify tests by knowledge of the code implementation, thus is the hypocrisy of me :)

Reply:
Here's an article that might explain it better than me, https://www.guru99.com/equivalence-partitioning-boundary-value-analysis.html
It's not actually "searching an iterable" so much as boundary checking the iterable with the addition of empty
```


        ┌────────┬────────┬────────┬────────┐
        │        │        │        │        │
        └────────┴────────┴────────┴────────┘
      ▲     ▲                 ▲         ▲     ▲
      │     │                 │         │     │
```
The boundary is "before" the iterable, the first item, the last item, after the iterable, and then you want one from the middle for paritioning equivalence

Reply:
I think it's fine to check for these cases. We may consider implementing this as a standard elsewhere in the codebase as well. Done.